Отображение должностей в рации. Обновлено под апстрим

Requiem4soul opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
**Обновление отображение должностей в рации.** Старая версия: [PR 446](https://github.com/imperial-space/space-station14-public/pull/446)

## Why / Balance
Для баланса вселенной. **Обновление для старой версии**

## Technical details
**Всё ещё можно создать прототип карты, он будет в словаре. Максимально оптимизировано.**

![изображение](https://github.com/user-attachments/assets/4a71360f-3cbf-4f81-8b05-1af6ed9ae361)

Requiem4soul committed:
Отображений должностей в рации

Обновлено под апстрим + фиксы

Requiem4soul committed:
Забыл что надо поменять названия

Requiem4soul committed:
Важный фикс

Все должности с "-–—−" ломали вывод должности после форматирования, но могли использоваться в базовых должностях "Шеф повар". Теперь они будут заменяться на пробел

CREAsTIVE commented on Content.Server/Imperial/Radio/JobPlayerAndColor.cs:
```diff
@@ -0,0 +1,173 @@
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+
+    // Стреляй, О(1) не отдам
+    private static Dictionary<string, string> SaveColorForJob = new Dictionary<string, string>();
```
Если уж и кешируешь значения прототипов, то храни хотя бы `Dictionary<string, ProtoId<IdColorAndJobPrototype>>`. А вообще в общем сообщении глянь другое решение.

Requiem4soul replied:
Пу-пу-пу, привет! Сейчас буду объяснять
`SaveColorForJob ` создан для хранения **уже вычисленных результатов (string)**. Это позволяет избавиться от повторных поисков уже найденных значений, что экономит производительность и считается важным в сетевых приложениях, где важна оптимизация. (мемоизация в программировании)
Так же использование строк в качестве ключей и значений минимизирует затраты ресурсов. При добавлении абстракции `ProtoId` увеличиться сложность и может повлиять на время выполнения операции, в чём и была изначальная задумка хэш таблица с мемоизацией результатов.

Крайне ошибочно было считать что тут хранятся прототипы. **Данный словарь отвечает именно за кэширование уже найденных результатов для повторном использовании данных. У него совершенно иная функция**

CREAsTIVE replied:
Вот только сами офицалы **настоятельно** не рекомендуют заниматься ручным кешированием. Данная разница в производительности (МИНИМАЛЬНАЯ прошу заметить) не стоит перед "чистотой кода". + Я предложил такой же по скорости вариант, но без ручного кеширования, см основной коммент

CREAsTIVE commented on Content.Server/Imperial/Radio/JobPlayerAndColor.cs:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    // Словарь для перевода должностей и для должностей состоящих не только из букв
+    private static readonly Dictionary<string, string> JobTranslations = new()
```
Долой хардкод

Requiem4soul replied:
Да нет, ты чего. Мемоизация - это база в динамическом и сетевом программировании. Буквально оптимизация. Скорее всего ты просто не совсем понял для чего это нужно, выше описал

CREAsTIVE commented on Content.Server/Imperial/Radio/JobPlayerAndColor.cs:
```diff
@@ -0,0 +1,173 @@
+    public void LoadPrototypes()
+    {
+        // Получаем все прототипы для idColor
+        var prototypes = IoCManager.Resolve<IPrototypeManager>().EnumeratePrototypes<IdColorAndJobPrototype>();
```
```suggestion
        var prototypes = _prototypeManager.EnumeratePrototypes<IdColorAndJobPrototype>();
```

Requiem4soul replied:
Выше объяснял что словарь отвечает за мемоизацию уже рассчитанных результатов. Иная функция.

Закрываю, так как выше объяснял

CREAsTIVE replied:
Не понял, тут я предлагаю не ресолвить заного прототайп менеджер, а использовать депенденси. В чём тут мемоизация?

CREAsTIVE commented on Content.Server/Imperial/Radio/JobPlayerAndColor.cs:
```diff
@@ -0,0 +1,173 @@
+                // Делаем начало должности с заглавной буквы и сохраняем в playerJob
+                playerJob = char.ToUpper(playerJob[0]) + playerJob.Substring(1);
+                // Заменяем тире, дефисы и минусы на пробелы. Эти символы при форматировании ломают вывод сообщения в рацию
+                playerJob = Regex.Replace(playerJob, @"[-–—−]", " ");
```
Регулярки для таких задач в целом... meh, но если и используешь их, то инициализируй хотя бы паттерн статически:
```
static Regex NameFixer = new(@"[-–—−]")
...
NameFixer.Replace(playerJob, " ")
```

Requiem4soul replied:
Обрабатывать входные результаты в данном случае обязательно, так как при форматировании текста функцией офов, такие значения будут ломать вывод в целом. Так же данное часто используется для защиты от SQL инъекций и вполне себе нормальная практика. 

По поводу почему тут как раз не стоит использовать `static`:
Регулярное выражение в данном месте используется в исключительно единственном методе и только при условии найденной работы. Для других значений оно не нужно. Создание `static` тут лишь увеличит сложность кода (О - которая)
Использование `static` тут имело смысл только если оно используется в нескольких местах или вызывается многократно раз. Поэтому дополнительное усложнение не оправдано.
Так же напомню что компилятор уже оптимизирует локально инициализированные строки.

Объяснил, поэтому закрываю

CREAsTIVE replied:
Как раз таки такое компилятор вполне может и пропустить. В шарпе нет понятий конст-експр конструкторов. Проблема в том, что ты инициализируешь каждый раз новую регулярку из строки, а значит дерево регулярки тут как раз таки строится каждый раз заново (если конечно шарп компилятор этого не заметил, что вполне вероятно)

CREAsTIVE commented on Content.Server/Imperial/Radio/JobPlayerAndColor.cs:
```diff
@@ -0,0 +1,173 @@
+    public string? GetColorPlayer(string? jobPlayer)
+    {
+        // Проверка. Работаем только тогда, когда работа была определена успешно
+        if (jobPlayer != null)
```
Инверсируй условия:
```suggestion
        if (jobPlayer is null) return "lime"; // Работа не определена
```

Requiem4soul replied:
Не, креатив, я настоятельно тебя попрошу вникнуть в то, что делает код. Ты буквально просишь создать ошибку уже.
`!=` это наоборот не равно `null`. У тебя `is` будет аналогом `==`

**Креатив, я прошу взглянуть на код внимательно и понять принцип его работы. Нельзя предлагать такие изменения, которые наоборот ломают код. Будь пожалуйста внимателен** 

Закрываю с пометкой ### "НИ В КОЕМ СЛУЧАЕ НЕЛЬЗЯ ТАК ДЕЛАТЬ!"

CREAsTIVE replied:
Посмотри внимательно на чендж реквест: я перенёс `return` в это условие

А я и не отрицаю. Инверс условий это техника, что позволяет сильно сократить вложенность. 
В твоём коде:

```cs
if (validate) {
  doSomething();
  if (validate) {
    doMore();
  }
  else {
    errorHandle();
  }
}
else {
  errorHandle();
}
```

Когда я настоятельно рекомендую:

```
if (!validate) errorHandle();
doSomething();
if (!validate) errorHandle();
doMore();
```

CREAsTIVE commented on Content.Server/Imperial/Radio/JobPlayerAndColor.cs:
```diff
@@ -0,0 +1,173 @@
+    /// </summary>
+    /// <param name="jobPlayer"> Нужно получить из метода GetJobPlayer. Или же вставляйте сюда свою string? работу, но нужно привести её к нижниму регистру для словаря. Метод ваш_стринг.ToLower()</param>
+    /// <returns></returns>
+    public string? GetColorPlayer(string? jobPlayer)
```
Не вижу ни одной причины возвращать нуллейбл. По идее этот метод должен уже точно определять цвет профессии, что бы не было лишних null проверок
```suggestion
    public string GetColorPlayer(string? jobPlayer)
```

Requiem4soul replied:
Мышка в рацию сказала сквик, у неё функция по поиску jobTitle через pda выдаст null, так как у мыши её нет

Не использовать тут string? - будет создавать фатальные ошибки...

### Креатив, очень прошу быть вдумчивым в ревью, чтобы не предлагать людям то, что сломает наоборот код

Закрываю с меткой ### "НИ В КОЕМ СЛУЧАЕ ТАК НЕ ДЕЛАТЬ!"

CREAsTIVE replied:
Мышка в рацию сказала сквик, и эта функция должна вернуть цвет игрока (исходя из названия профессии), а не *возможно возвращать* цвет игрока. Если мышка сказала сквик, то пусть уж эта функция возвращает как раз таки дефолтный цвет (что я так же отмечал в ревью)

CREAsTIVE commented on Content.Server/Imperial/Radio/JobPlayerAndColor.cs:
```diff
@@ -0,0 +1,173 @@
+            IsDictionaryEmpty();
+
+            // Проверка наличия должности в словаре
+            if (SaveColorForJob.ContainsKey(normalizedJob))
+            {
+                // Если должность найдена в словаре, возвращаем соответствующий цвет
+                string color = SaveColorForJob[normalizedJob];
+                return color;
+            }
+            else
+            {
+                // Если должность не найдена в словаре, возвращаем дефолтный цвет
+                return "lime";
+            }
```
Инвёрс условия:
```suggestion
            // Если должность не найдена
            if (!SaveColorForJob.ContainsKey(normalizedJob)) return "lime";
            
            // Возвращаем цвет должности
            string color = SaveColorForJob[normalizedJob];
            return color;
```

(А вообще лучше сразу избавиться от двойного поиска:
```suggestion
return SaveColorForJob.TryGetValue(normalizedJob, out var color) ? color : "lime"
```
)

Requiem4soul replied:
Отсутствие обработки ошибок будет приводить к тому, что ошибки не будут обрабатываться. Так же в jobPlayer мы уже задаём значения для тех работ, которые невозможно найти или их нет

Закрываю

CREAsTIVE replied:
Мой изменённый код полностью идентичен твоему коду и работает точно так же, НО не совершает поиск по словарю дважды (что как раз таки и вредит оптимизации)

CREAsTIVE commented on Content.Server/Imperial/Radio/JobPlayerAndColor.cs:
```diff
@@ -0,0 +1,173 @@
+            else
+            {
+                // Если должность не найдена в словаре, возвращаем дефолтный цвет
+                return "lime";
```
Дефолтный цвет стоит вынести в константу

Requiem4soul replied:
это дефолтный цвет, используемый исключительно в этом конкретном методе. Его значение понятно из контекста и не используется повторно в других частях проекта. Неоправданный вынос константы для единоразового использования в методе.

Закрываю

CREAsTIVE commented on Content.Server/Imperial/Radio/JobPlayerAndColor.cs:
```diff
@@ -0,0 +1,173 @@
+        }
+
+        // Определяем какой цвет должен быть у данной должности
+        string? nameColorPlayer = GetColorPlayer(nameJobPlayer);
```
```suggestion
        string nameColorPlayer = GetColorPlayer(nameJobPlayer);
```

Requiem4soul replied:
Объяснял выше, с пометкой ### "НИ В КОЕМ СЛУЧАЕ ТАК НЕ ДЕЛАТЬ!"

CREAsTIVE replied:
Объяснил выше)

CREAsTIVE commented on Content.Server/Imperial/Radio/JobPlayerAndColor.cs:
```diff
@@ -0,0 +1,173 @@
+        string? nameColorPlayer = GetColorPlayer(nameJobPlayer);
+
+        // Если всё сработало правильно, будет работать этот код. Я даже не могу придумать случаев, когда будет работать не это условие, а return вконце метода
+        if ((nameColorPlayer != null) && (nameJobPlayer != null))
```
Убрать условия, т. к. эти поля более не могут быть nullable

Requiem4soul replied:
Данные проверки являются стандартными и активно используются офами в их оригинальном коде. Данные проверки необходимы для обработки ошибок и багов.

Пример из ChatSystem метод TrySendInGameICMessage:
`if (HasComp<GhostComponent>(source))
        {
            // Ghosts can only send dead chat messages, so we'll forward it to InGame OOC.
            TrySendInGameOOCMessage(source, message, InGameOOCChatType.Dead, range == ChatTransmitRange.HideChat, shell, player);
            return;
        }

        if (player != null && !_chatManager.HandleRateLimit(player))
            return;

        // Sus
        if (player?.AttachedEntity is { Valid: true } entity && source != entity)
        {
            return;
        }

        if (!CanSendInGame(message, shell, player))
            return;`
            
Закрываю

CREAsTIVE replied:
Данное изменение предложено на основе того, что мы отказываемся от nullable строк. В твоём же примере немного другая ситуация.

CREAsTIVE commented on Content.Server/Radio/EntitySystems/RadioSystem.cs:
```diff
@@ -1,3 +1,4 @@
+using Content.Server.Imperial.Radio;
```
```suggestion
using Content.Server.Imperial.Radio; // Imperial RadioPlayerAndJob
```

Requiem4soul replied:
Было показано на разработке психоза, что using и зависимости у нас (как и у других) не помечаются лицензией, так как лицензия вешается на саму реализацию, а не на использование какого-то метода

CREAsTIVE replied:
Я так же думал, однако меня позже миртон просил даже юзинги помечать меткой империала. Увы

CREAsTIVE commented on Content.Server/Radio/EntitySystems/RadioSystem.cs:
```diff
@@ -28,6 +29,7 @@ public sealed class RadioSystem : EntitySystem
     [Dependency] private readonly IPrototypeManager _prototype = default!;
     [Dependency] private readonly IRobustRandom _random = default!;
     [Dependency] private readonly ChatSystem _chat = default!;
+    [Dependency] private readonly JobPlayerAndColor _jobPlayer = default!;
```
```suggestion
    [Dependency] private readonly JobPlayerAndColor _jobPlayer = default!; // Imperial RadioPlayerAndJob
```

Requiem4soul replied:
В зависимостях и using мы не указываем лицензию на код.
Пример с системы психоза:
`using Content.Shared.Imperial.ICCVar;`
`public sealed class MassPsychosisRule : StationEventSystem<MassPsychosisRuleComponent>
{
    [Dependency] private readonly PsychosisSystem _psychosis = default!;
    [Dependency] private readonly IRobustRandom _random = default!;
    [Dependency] private readonly IConfigurationManager _cfg = default!;`
    
  Так, кажется на какой-то изменение не ответил верно и прикрепил как отдельное сообщение

CREAsTIVE commented on Content.Server/Imperial/Radio/JobPlayerAndColor.cs:
```diff
@@ -0,0 +1,173 @@
+    /// </summary>
+    /// <param name="uid"> uid игрока, у которого мы ищем карту</param>
+    /// <returns></returns>
+    public string? GetJobPlayer(EntityUid uid)
```
Нуллейбл и так не возвращается
```suggestion
    public string GetJobPlayer(EntityUid uid)
```

Requiem4soul replied:
Опять же вызов критической ошибки. Выше объяснялось. Закрываю с пометкой ### "НИ В КОЕМ СЛУЧАЕ ТАК НЕ ДЕЛАТЬ!"

CREAsTIVE commented on Content.Server/Imperial/Radio/JobPlayerAndColor.cs:
```diff
@@ -0,0 +1,173 @@
+    public string? GetJobPlayer(EntityUid uid)
+    {
+        // Проверяем нашлась ли карта и какое id у карты
+        if (_idCardSystem.TryFindIdCard(uid, out var id))
```
Инверсия условий!!!

Requiem4soul replied:
Данное ясно выражает намерение: "Если ID-карта найдена, выполняем дальнейшие действия". Инверсия этого условия приведёт к добавлению лишнего отрицания и увеличению ненужной нагрузки.

Мы не можем позволить себе тут сделать ранний выход, так как задача стоит в том, что для всех не найденных id и у тех, кто её в целом иметь не может, должно отображаться должность "Неизвестно"

Закрываю

CREAsTIVE commented on Content.Server/Imperial/Radio/JobPlayerAndColor.cs:
```diff
@@ -0,0 +1,173 @@
+            // Мы сохраняем работу
+            string? playerJob = id.Comp.LocalizedJobTitle;
+            // Если работа нашлась верно, мы начинаем основной процесс
+            if (playerJob != null)
```
Инверсия условий!!

Requiem4soul replied:
Аналогично выше, закрываю, так как нам необходимо обрабатывать такие исключения

CREAsTIVE commented on Content.Server/Imperial/Radio/JobPlayerAndColor.cs:
```diff
@@ -0,0 +1,173 @@
+    public string CompletedJobAndPlayer(EntityUid uid,  string? name)
+    {
+        // Активируем метод для определения должности
+        string? nameJobPlayer = GetJobPlayer(uid);
```
```suggestion
        string nameJobPlayer = GetJobPlayer(uid);
```

Requiem4soul replied:
Опять предложение критической ошибки. Закрываю с пометкой ### "НИ В КОЕМ СЛУЧАЕ ТАК НЕ ДЕЛАТЬ!"

Закрываю, выше объяснено почему такое будет приводить к критической ошибке

CREAsTIVE commented on Content.Server/Imperial/Radio/JobPlayerAndColor.cs:
```diff
@@ -0,0 +1,173 @@
+        // Активируем метод для определения должности
+        string? nameJobPlayer = GetJobPlayer(uid);
+        // Если должность определена, добавляем
+        if (nameJobPlayer != null)
```
Более не требуется

Requiem4soul replied:
Закрываю, так как выше объяснял

CREAsTIVE reviewed (changes requested):
В следующий раз пожалуйста создавай отдельную ветку.

На текущий момент у тебя система полностью зависит от локализации. Я предлагаю такой вариант:

В yml хранится что то такое:
```yml
- type: customJobColor
  id: Detective # Такое же, как и id профессии
  color: red
  
- type: customJobColor
  id: CentralCommandOfficial
  color: blue
  bold: true # Доп параметры опционально
```

После в коде просто получаем id профессии и через `_protoManager.Index<CustomJobColor>(jobId)` (что так же твой любимый O(1)) получаем всю необходимую информацию о форматировании профессии.

Тогда никакого ручного кеширования делать будет не надо, никакой зависимости от локализации тоже не будет, и можно будет легко вводить новые параметры.

Есть ещё вариант в модификации существующего кода профессий и внесения туда дополнительных полей. Но боюсь миртону не понравится мерджить такое с апстримом

Requiem4soul commented:
> В следующий раз пожалуйста создавай отдельную ветку.
> 
> На текущий момент у тебя система полностью зависит от локализации. Я предлагаю такой вариант:
> 
> В yml хранится что то такое:
> 
> ```yaml
> - type: customJobColor
>   id: Detective # Такое же, как и id профессии
>   color: red
>   
> - type: customJobColor
>   id: CentralCommandOfficial
>   color: blue
>   bold: true # Доп параметры опционально
> ```
> 
> После в коде просто получаем id профессии и через `_protoManager.Index<CustomJobColor>(jobId)` (что так же твой любимый O(1)) получаем всю необходимую информацию о форматировании профессии.
> 
> Тогда никакого ручного кеширования делать будет не надо, никакой зависимости от локализации тоже не будет, и можно будет легко вводить новые параметры.
> 
> Есть ещё вариант в модификации существующего кода профессий и внесения туда дополнительных полей. Но боюсь миртону не понравится мерджить такое с апстримом

### Креатив, стоит относиться к ревью более обдуманно и внимательнее!
 Ладно одно дело ты не понял для чего нужен хэш словарь, который отвечает за мемоизацию уже вычисленных результатов, со всеми бывает, к критической ошибке это не приведёт
 
 ### Но делать значение jobPlayer не string? - когда у тебя буквально она в некоторых случаях будет иметь значение null - это то, что ни в коем случае делать нельзя.
 
 Теперь даже если в ревью ты будешь людям предлагать изменения, они будут просто "на всякий случай" не вводить такие изменения, так как в каком-то ревью **ты предлагал такое изменение, которое бы приводило к критическим ошибкам.** Так делать не стоит. **К ревью стоит относится внимательно и аккуратно!**

Requiem4soul commented:
### Изменения от Креатива, в связи с просьбой созданием критических ошибок и удалением оптимизации было целиком отклонено.
Ни одна критическая ошибка в коде не была принята, оптимизация так же сохранена. **Код не изменился**

CREAsTIVE commented:
> ### Изменения от Креатива, в связи с просьбой созданием критических ошибок и удалением оптимизации было целиком отклонено.
> Ни одна критическая ошибка в коде не была принята, оптимизация так же сохранена. **Код не изменился**

Ты не понял ни единого моего изменения. Зайди в ДС по возможности я объясню

Requiem4soul commented:
> > ### Изменения от Креатива, в связи с просьбой созданием критических ошибок и удалением оптимизации было целиком отклонено.
> > Ни одна критическая ошибка в коде не была принята, оптимизация так же сохранена. **Код не изменился**
> 
> Ты не понял ни единого моего изменения. Зайди в ДС по возможности я объясню

Креатив, за этот код **я отвечаю головой.** Я не могу осознанно вводить то, **что приведёт к фатальной ошибке**
![изображение](https://github.com/user-attachments/assets/0a041ad4-ed8d-4f33-aa6f-621bdb56e6df)


Попробуй просто сам сделать то, что ты предложил. Затем попробуй снять с pda и не иметь в руках карту, и что-то сказать в рацию. При определении работы метод офов вернёт `null`, **который у тебя не обрабатывается**.

**UPD:** Для того чтобы проверить на ошибку код Креатива - проще или через карту агента оставить строку "Должность" пустой, или через консоль ГП, или через компоненты сделать пустое поле. **Сообщение не появится в чате**, **отобразиться просто критическая ошибка (красным)**, потом входит в какой-то странный цикл с попыткой отправится, и **лишь спустя время появляется фатальная ошибка**
![изображение](https://github.com/user-attachments/assets/0630f934-2361-4a89-a72c-4a3966d9fed7)
![изображение](https://github.com/user-attachments/assets/b8dd7f89-1006-45e9-a918-88bb7b005fc1)



### Я не могу принять критическую ошибку ни при каких условиях.
Креатив, пойми, это не потому что ты какой-то такой не хороший. А потому что я не имею право осознанно ввести критическую ошибку в свой код, который могут принять и использовать

Ну не просто же так офы делают точно такую же обработку у себя. Не стоит считать что они просто так такое делают
![изображение](https://github.com/user-attachments/assets/fc794d11-5454-4420-9895-5936269c36e6)

CREAsTIVE commented:
@Requiem4soul Просто свяжись со мною через ДС, и я объясню. Данное изменение предполагает другие запрошенные изменения. Я готов при тебе все их применить и показать, что и как.

Requiem4soul commented:
@mirton14  Доброго Вам дня! Прошу прощения за пинг, но дело **реально важное**. 
В моём ПР(е) (Отображение должностей в рации. Обновлено под апстрим #49) товарищ Креатив в ревью предложил сделать код, который будет выдавать критическую ошибку вначале, затем фатальную, так как он исключил обработку null значений. Мною было это несколько раз указано Креативу (рассказывал почему приведёт, как проверить ошибку, делал скриншоты с указанием и видео с демонстрацией). Но до сих пор Креатив отказывается как-либо комментировать зачем он предложил мне в ревью критическую ошибку

Я бы ни слова не сказал, **не будь у товарища Креатива роль разработчика на данном сервере, что указывает на то, что он часть официальной команды разработчиков**. Хорошо я понимал хорошо свой код и сразу понял что такое допустить нельзя, но, если бы данное произошло с новичком, и такому новичку Креатив сказал сделать код, у которого будет критическая ошибка, которая крашнет сервера - **новичок мог бы необдуманно доверится официальному разработчику ИС, а потом нести ответственность за то, в чём он не виноват**.
Так же **это в целом убивает доверие к ревью, когда ты знаешь что другой разработчик может намеренно в ревью дать тебе код с критической ошибкой**. К здоровой атмосфере внутри разработчиков такое не приведёт

### Прошу вас разобраться в данной ситуации, так же выяснить у Креатива было это сделано намеренно или нет

*Видео материал и длительное обсуждение с Креативом было в данной ветке дискорда (https://discord.com/channels/1040938900039929917/1304511097578979329/1312203330155712604)

Qsan163 commented:
потому что понабирают всяких девелоперов

Requiem4soul commented:
> потому что понабирают всяких девелоперов

А у вас есть хотя бы какой-то собес на джуна? Или джунов по сути почти всем выдают, но прав у них нет?

Просто это же реально опасно, если набегаторы эволюционируют, станут всем как "разработчик ИСа" предлагать ввести фатальные ошибки, а потом их коллеги набегаторы буду крашить сервера при помощи такого способа. Реально опасная же вещь

CREAsTIVE commented:
> > потому что понабирают всяких девелоперов
> 
> А у вас есть хотя бы какой-то собес на джуна? Или джунов по сути почти всем выдают, но прав у них нет?
> 
> Просто это же реально опасно, если набегаторы эволюционируют, станут всем как "разработчик ИСа" предлагать ввести фатальные ошибки, а потом их коллеги набегаторы буду крашить сервера при помощи такого способа. Реально опасная же вещь

Вы настоятельно продолжаете игнорировать мой ПР в вашу репу, где **рабочий** код, с изменениями из ревью, который вы, конечно же, не готовы проверять

Requiem4soul committed:
Улучшена защита от креативных ошибок

1) Раньше ГМ мог по ошибке закодить прототип с значениями null, и такие значения бы попали в словарь. Я же исправил это, и теперь в словарь мы добавляем только значения без null
2) Улучшена защита от креативных ошибок, которые могли бы привести к крашу сервера (теперь на каждом этапе мы будем уверены что любое вводимое значение не приведёт к ошибке). Методы всё так же могут принять null, но превратить его в корректное значение. Если же каким-то образом из метода выйдет null, то главный метод сможет принять и обработать такие ошибки.
А именно: защита от имени с значением null, защита от должности с значением null, защита от цвета с значением null

Qsan163 commented:
> > потому что понабирают всяких девелоперов
> 
> А у вас есть хотя бы какой-то собес на джуна? Или джунов по сути почти всем выдают, но прав у них нет?
> 
> Просто это же реально опасно, если набегаторы эволюционируют, станут всем как "разработчик ИСа" предлагать ввести фатальные ошибки, а потом их коллеги набегаторы буду крашить сервера при помощи такого способа. Реально опасная же вещь

если я не ошибаюсь , то есть , изначально человек заполняет анкету , потом собес с HR - ом , устный или на крайняк онлайновый , а потом уже человек становится сотрудником

Qsan163 commented:
я слышал у нас нехватка кадров в девелоперной , возможно по этому такие есть

Requiem4soul committed:
Лицензия MIT с авторством

mirton14 commented:
Не считаю необходимым иметь конкретно эту фичу в билде.
